proxy/src/http/conn_pool.rs

```diff
@@ -157,25 +165,33 @@ impl GlobalConnPool {
         Ok(())
     }
 
+    /// find or create a pool for this endpoint.
     async fn get_endpoint_pool(&self, endpoint: &String) -> Arc<Mutex<EndpointConnPool>> {
-        // find or create a pool for this endpoint
         let mut created = false;
-        let mut global_pool = self.global_pool.lock();
-        let pool = global_pool
-            .entry(endpoint.clone())
-            .or_insert_with(|| {
-                created = true;
-                Arc::new(Mutex::new(EndpointConnPool {
-                    pools: HashMap::new(),
-                    total_conns: 0,
-                }))
-            })
-            .clone();
-        let global_pool_size = global_pool.len();
-        drop(global_pool);
+        // try a read only lock first.
+        let pool = match self.global_pool.get(endpoint) {
+            Some(pool) => pool.clone(),
+            None => {
+                // fall back to more expensive write lock
+                self.global_pool
+                    .entry(endpoint.clone())
```

Comment:
you already checked that it’s not present in hashmap. Why not to use usual `insert`?

Reply:
Race conditions. It could technically be inserted between me checking and actually inserting